Content/RootScript/AngelEngine/GUI/GUI.as

```diff
@@ -0,0 +1,597 @@
+	}
+
+	// # Monitor the mouse for clicks.
+	shared void updateMouse() {
+		// bool is1hit=Input::Mouse1::isHit(); // This is not necessary.
+		bool is1down=Input::Mouse1::isDown();
+		if(is1down) { if(!isMouse1Down) { onClick(); isMouse1Down=true; } }
+		else if (!is1down && isMouse1Down) { isMouse1Down=false; }
+	}
+
+
+	// #### Text Entry ----
+	// Handles textentry focus and the input capturer.
+	// Mostly used and managed by GUI::TextEntry instances.
+	shared GUITextEntry@ textEntryFocus;
+	shared bool validTextEntering() { if(@textEntryFocus!=null) { if(!textEntryFocus.visible) { @textEntryFocus=null; return false; } return true; } return false; }
+	shared void startTextEntering(GUITextEntry@ gui) { @textEntryFocus=@gui; Input::startTextInputCapture(); }
+	shared void stopTextEntering() { @textEntryFocus=null; Input::stopTextInputCapture(); }
+	shared void updateTextEntering() { if(validTextEntering()) { textEntryFocus.updateTextEntering(); } }
+	shared void clickedTextEntering(Vector2f mpos) {
+		if(validTextEntering() && !pointInSquare(mpos,textEntryFocus.paintPos,textEntryFocus.paintSize)) { stopTextEntering(); }
+	}
+	shared funcdef void TextEnteredFunc(string&in input);
+
+
+	// #### Cascade GUI Renderer & Updater.
+	shared float interp;
+	shared void startRender(float interp) {
+		GUI::interp=interp;
+		UI::setTextureless();
+		UI::setColor(Color::White);
+		for(int i=0; i<baseInstances.length(); i++) { if(baseInstances[i].visible) { baseInstances[i].drillPreRender(); } }
+		executeLayout();
+		for(int i=0; i<baseInstances.length(); i++) { if(baseInstances[i].visible) { baseInstances[i].drillRender(); } }
+		UI::setTextureless();
+		UI::setColor(Color::White);
+	}
+
+	shared void startUpdate(float interp) {
+		GUI::interp=interp;
+		updateMouse();
+		updateTextEntering();
+		int baseLen=baseInstances.length();
+		for(int i=0; i<baseLen; i++) { if(baseInstances[i].visible) { baseInstances[i].drillUpdate(); } }
+
+	}
+
+	shared void initialize() {
+		updateResolution();
+	}	
+}
+
+
+
+
+// # GUI Base Class --------
+// The base GUI class on which all GUI classes are based.
+// On its own, it acts like a container for other GUI elements.
+// All GUI elements are derived from this class with overrides and constructors to give behaviour and purpose.
+// Created with GUI@ gui_panel = GUI();
+// To create a panel parented to another, simply use GUI@ child_panel = GUI(@parent_panel);
+// Then configure the child, such as .align=GUI::Align::Fill.
+// To manage layouts, you must call invalidateLayout() manually as-needed (in the majority of cases), as this is rarely called automatically.
+// layout updates are costly, requiring repositioning and resizing of all children elements, but the elements do not change often so this is worthwhile.
+// And sometimes you must change layout data without triggering an update, so having no automatic updates helps with that.
+
+shared class GUI {
```

Comment:
This class is clearly missing access modifiers for a lot of its members. When possible, make members private or protected.

Reply:
Confirmed, I felt the same way while writing it. This ought to be done in the next iteration of the GUI system if we are happy to continue with this design.